Adding docs for migrating from SDKv2 to the Framework (#461)
hashicorp/terraform-plugin-framework: website/docs/plugin/framework/migrating/providers/index.mdx

---
page_title: 'Provider: Migrating from SDKv2 to the Framework'
description: >-
  Migrate a provider definition and schema from SDKv2 to the plugin Framework.
---

# Provider

Providers are Terraform plugins that define resources and data sources for practitioners to use. You serve your
providers with a provider server so they can interact with Terraform.

This page explains how to migrate a provider server, definition, and schema from SDKv2 to the plugin Framework.

## Serving the Provider

You must update your provider's `main.go` file to serve Framework providers. Refer to [Provider Servers](/plugin/framework/provider-servers) in the Framework documentation for details.

### SDKv2

> In SDKv2, the provider packages `main` function serves the provider by calling `plugin.Serve`.

Reviewer: ```suggestion
In SDKv2, the provider package's `main` function serves the provider by calling `plugin.Serve`.
```

    return map[string]provider.DataSourceType{
        "tls_public_key":  &publicKeyDataSourceType{},
        /* ... */
    }, nil
}

func (p *provider) GetSchema(_ context.Context) (tfsdk.Schema, diag.Diagnostics) {
    return tfsdk.Schema{
        Attributes: map[string]tfsdk.Attribute{
            "proxy": {
            /* ... */
            },
        },
    }, nil
}

func (p *provider) Configure(ctx context.Context, req provider.ConfigureRequest, res *provider.ConfigureResponse) {
    /* ... */
}
```

## Provider Schema

A provider schema defines the attributes and behaviors of the provider itself. For example, a provider that connects to
a third-party API may define attributes for the base URL or a required authentication token.

### SDKv2

In SDKv2, you implement a provider Schema by populating the `Schema` field on the `schema.Provider` struct. The `Schema`
field contains a `map[string]*schema.Schema`. Each map entry represents the name of the attribute and pointer to a
`schema.Schema` struct that defines that attribute's behavior.

The following example defines the provider schema in the `Schema` field within the `schema.Provider` struct.

```go

> func New() *schema.Provider {
>     return &schema.Provider{
>         Schema: map[string]*schema.Schema{},
>         /* ... */

Reviewer: ```suggestion
func New() *schema.Provider {
    return &schema.Provider{
        Schema: map[string]*schema.Schema{
            /* ... */
        },
```
Just to re-enforce that the stuff mentioned above goes inside the map definition.

> func New() *schema.Provider {
>     return &schema.Provider{
>         Schema: map[string]*schema.Schema{},
>         /* ... */
> ```

Reviewer: ```suggestion
func New() *schema.Provider {
    return &schema.Provider{
        Schema: map[string]*schema.Schema{
        /* ... */
        },
```

Just to re-enforce that "the stuff" goes inside the brackets.

### Framework

In the Framework, the `GetSchema` function returns the provider schema. The `GetSchema` function is part of the
`provider.Provider` interface that your provider must implement. `GetSchema` returns a struct containing fields for
`Attributes` and `Blocks`. These `Attributes` and `Blocks` contain `map[string]tfsdk.Attribute` and
`map[string]tfsdk.Block`, respectively. Refer to [Providers - GetSchema](/plugin/framework/providers#getschema) in the
Framework documentation for details.

The following code shows the `GetSchema` function, which returns the provider schema.

```go

> func (p *provider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
>     return tfsdk.Schema{}, nil
> }

Reviewer: ```suggestion
func (p *provider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
    return tfsdk.Schema{
        /* ... */
    }, nil
}
```
Same goal as previous comment - show where the stuff we're talking about goes.

> ```go
> func (p *provider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
>     return tfsdk.Schema{}, nil
> }
> ```

Reviewer: ```suggestion
```go
func (p *provider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
    return tfsdk.Schema{
            /* ... */
    }, nil
}
```
Same goal as prior suggestion - show where the stuff goes.

Refer to the [Attributes](/plugin/framework/migrating/attributes-blocks#attributes) and
[Blocks](/plugin/framework/migrating/attributes-blocks#blocks) pages in this migration guide to learn how to migrate
those fields to the Framework.

### Migration Notes

Remember the following differences between SDKv2 and the Framework when completing the migration.

> In SDKv2, `schema.Schema` is a struct that defines attributes and behaviors (e.g., `Type`, `Optional`). In the
> Framework `tfsdk.Schema` is a struct that defines `Attributes`, which are structs that define attributes and behaviors
> (e.g., `Type`, `Optional`).

Reviewer: ```suggestion
- In SDKv2, `schema.Schema` is a struct that defines attributes and behaviors (e.g., `Type`, `Optional`). In the
Framework `tfsdk.Schema` is a struct that includes `tfsdk.Attributes`, which are structs that define attributes and behaviors
(e.g., `Type`, `Optional`).
```

Three small changes:
- Kinda weird to have a list with one bullet, but this matches the pattern in the rest of the guide.
- I think "includes" is more correct than "defines"?
- If we're going to use the full name for the other structs, do so for `tfsdk.Attributes` too.

### Example

The following examples show how to migrate portions of the [tls](https://github.com/hashicorp/terraform-provider-tls)
provider.

For a complete example, clone the
`terraform-provider-tls` repository and compare `provider.go` in
[v3.4.0](https://github.com/hashicorp/terraform-provider-tls/blob/v3.4.0/internal/provider/provider.go)
with [v4.0.1](https://github.com/hashicorp/terraform-provider-tls/blob/v4.0.1/internal/provider/provider.go).

This example also shows how to use a nested block and a nested attribute for the SDKv2 and Framework examples,
respectively. Refer to the
[Blocks and Nested Attributes](/plugin/framework/migrating/attributes-blocks) page in this guide for more details.

#### SDKv2

> The following example shows how to configure an attribute within a provider schema using SDKv2.

Reviewer: ```suggestion
The following example from the `provider.go` file shows the configuration of the `url` attribute for the  provider's`proxy` configuration block.
```

Match the way we introduce examples in other pages.

```go
Schema: map[string]*schema.Schema{
    "proxy": {
        Type:     schema.TypeList,
        Optional: true,
        MaxItems: 1,
        Elem: &schema.Resource{
            Schema: map[string]*schema.Schema{
                "url": {
                    Type:             schema.TypeString,
                    Optional:         true,
                    ValidateDiagFunc: validation.ToDiagFunc(validation.IsURLWithScheme(SupportedProxySchemesStr())),
                    ConflictsWith:    []string{"proxy.0.from_env"},
                    Description: "URL used to connect to the Proxy. " +
                        fmt.Sprintf("Accepted schemes are: `%s`. ", strings.Join(SupportedProxySchemesStr(), "`, `")),
                },
                /* ... */
```

#### Framework

> The following example shows how to configure provider schema using the Framework.

Reviewer: ```suggestion
The following shows the same section of provider code after the migration.

This code implements the `url` attribute for the `proxy` block with the Framework.
```

Match the way we talk about examples elsewhere.
